src/metabase/query_processor/parameters.clj

```diff
@@ -30,83 +128,83 @@
+    :filter (fn [{:keys [date-1 date-2]} field]
+              ["BETWEEN" field (date->iso8601 date-1) (date->iso8601 date-2)])}
+   ;; before day
+   {:parser (regex->parser #"~([0-9-T:]+)" [:date])
+    :range  (fn [{:keys [date]} _]
+              {:end date})
+    :filter (fn [{:keys [date]} field]
+              ["<" field (date->iso8601 date)])}
+   ;; after day
+   {:parser (regex->parser #"([0-9-T:]+)~" [:date])
+    :range  (fn [{:keys [date]} _]
+              {:start date})
+    :filter (fn [{:keys [date]} field]
+              [">" field (date->iso8601 date)])}
+])
 
-;; NOTE: this is perhaps a little hacky, but we are assuming that `dt` will be in the first month of the quarter
-(defn- quarter-range [^DateTime dt]
-  {:end   (t/last-day-of-the-month (t/plus dt (t/months 2)))
-   :start (t/first-day-of-the-month dt)})
-
-(defn- year-range [^DateTime dt]
-  {:end   (t/last-day-of-the-month  (.withMonthOfYear dt DateTimeConstants/DECEMBER))
-   :start (t/first-day-of-the-month (.withMonthOfYear dt DateTimeConstants/JANUARY))})
-
-(defn- absolute-date->range
-  "Take a given string description of an absolute date range and return a MAP with a given `:start` and `:end`.
-
-   Supported formats:
-
-      \"2014-05-10~2014-05-16\"
-      \"Q1-2016\"
-      \"2016-04\"
-      \"2016-04-12\""
-  [value]
-  (if (s/includes? value "~")
-    ;; these values are already expected to be iso8601 strings, so we are done
-    (zipmap [:start :end] (s/split value #"~" 2))
-    ;; these cases represent fixed date ranges, but we need to calculate start/end still
-    (->> (cond
-           ;; quarter-year (Q1-2016)
-           (s/starts-with? value "Q") (let [[quarter year] (s/split value #"-" 2)]
-                                        (quarter-range (start-of-quarter quarter (Integer/parseInt year))))
-           ;; year-month (2016-04)
-           (= (count value) 7)        (month-range (tf/parse (tf/formatters :year-month) value))
-           ;; default is to assume a single day (2016-04-18).  we still parse just to validate.
-           :else                      (let [dt (tf/parse (tf/formatters :year-month-day) value)]
-                                        {:start dt, :end dt}))
-         (m/map-vals (partial tf/unparse (tf/formatters :year-month-day))))))
-
-
-(defn- relative-date->range
-  "Take a given string description of a relative date range such as 'lastmonth' and return a MAP with a given
-   `:start` and `:end` as iso8601 string formatted dates.  Values should be appropriate for the given REPORT-TIMEZONE."
-  [value report-timezone]
+;; Collects all parameter value parsers and range/filter creators together
+(def ^:private date-param-values
+  (concat relative-date-param-values absolute-date-param-values))
+
+(defn- find-and-execute-date-param-method
+  [date-param-value method-id method-param]
+  (some (fn [{parser :parser method method-id}]
+          (if-let [parser-result (parser date-param-value)]
```

Comment:
if there's no "else" clause in Lisp languages you use `when` and `when-let` instead of `if` and `if-let`